README.md
# Romanization for Amazon Music α (RfAM) 🎵
By using the plugin you agree to share your IP, and the text of your music played, and you also agree that we will store the log of your requests for an indefinite period of time.

Ao usar o plugin você concorda em partilhar seu IP, e o texto da sua musica reproduzida, concordando também em armazenar os logs de seus pedidos por um tempo indeterminado

# 🇺🇸

!NOW SUPPORTS SPOTIFY!

Chrome Plugin to add version transliterate (in lantin) in japanese music with letter on Amazon Music web player

"Wow how cool... lyrics on the streaming service... i wanted to sing! but i don't know anything about nipponic writing..." YOUR PROBLEMS ARE OVER, at least on Amazon Music!

This open source plugin for the chromumium browser provides a romanized (or transliterated into Latin, call it what you will) version of Japanese songs.

Basic installation guide
----------------------------------
1. Download the [.CRX](https://github.com/Ruaneri-Portela/Romanization-For-Amazon-Music/raw/main/Extension.crx) or [.ZIP](https://github.com/Ruaneri-Portela/Romanization-For-Amazon-Music/raw/main/Extension.zip), if you download the ZIP unzip it into a folder that contains only the extension files after decompression
2. Go to your extension page in your favorite browser
3. Put it in dev mode.
4. Load the extension, in the files has the .crx, but for example the EDGE only has the option to load the unzipped version. You will select the folder in case you load the unpacked extension.
5. After that the extension should be working, test on Amazon Music 😊

Requirements... 
----------------------------------
Client:
Plugin and constant internet connection.

Server (Read about the API in end this document):
Python, Flash, Gunicorn and KAKASI

Kakasi is a transliteration software in C, it is used as our backend to convert the hiragana, katakana and kanji into romanji... More information at http://kakasi.namazu.org/index.html.en or https://github.com/loretoparisi/kakasi

# 🇧🇷

! AGORA SUPORTAMOS O SPOTIFY !

"Nossa que maneiro... letras no serviço de streaming ... queria cantar! mas não sei nada da escrita nipponica..." SEUS PROBLEMAS ACABARAM, ao menos no Amazon Music!

Esse plugin de codigo aberto para o chromumium browser disponibiliza uma versão romanizada (ou transliterada para o latin, chame como quiser) de musicas em japonês.

Guia Básico de instalação
----------------------------------
1. Baixe o [.CRX](https://github.com/Ruaneri-Portela/Romanization-For-Amazon-Music/raw/main/Extension.crx) ou o [.ZIP](https://github.com/Ruaneri-Portela/Romanization-For-Amazon-Music/raw/main/Extension.zip), caso baixe o ZIP descompacte dele em uma pasta que contenha apenas os arquivos da extensão apôs a descompressão
2. Vai em sua pagina de extensão em seu navegador favortito
3. Coloque ele em modo dev.
4. Carrege a extenção, nos arquivos tem o .crx, mas por exemplo o EDGE só tem a opção de carregar a versão descompactada. Você vai selecionar a pasta no caso de carregar a extensão não empacotada.
5. Após isso a extensão deve ser está funcionando, teste no Amazon Music 😊
   
Requisitos...
----------------------------------
Cliente:
Plugin e conexão constante com a internet.

Servidor (Leia sobre a API no fim desse documento):
Python, Flash, Gunicorn e o KAKASI

Kakasi e um software de transliteração em C, ele e usado como nosso backend para converter os hiragana, katakana e kanjis em romanji... Mais informações em http://kakasi.namazu.org/index.html.en ou https://github.com/loretoparisi/kakasi


----------------------------------
Original
![Não Romanizado/Non Romanized](https://github.com/Ruaneri-Portela/Romanization-For-Amazon-Music/blob/main/github/non-romanized.png?raw=true)
Com o plugin/ At plugin
![Romanizado/Romanized](https://github.com/Ruaneri-Portela/Romanization-For-Amazon-Music/blob/main/github/romanized.png?raw=true)

[API informations HERE](https://github.com/Ruaneri-Portela/KakasiAPI-Docker)

[Informações da API AQUI](https://github.com/Ruaneri-Portela/KakasiAPI-Docker)


Changelog.txt
0.1 ---------------------------------------------------
Initial build

0.2 ---------------------------------------------------

Now support Spotify!

content-script.js
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
function romanize(dataIn) {
  const url = 'http://localhost:5000/convert';
  const inputData = 'input=' + dataIn;

  return fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded'
    },
    body: inputData
  })
    .then(response => response.text())
    .then(data => {
      return data;
    })
    .catch(error => {
      throw error;
    });
}
async function romanizeStrings() {
  let check = 0;
  let percursion = 0;
  let stringsArray;
  let ArrayCheck;
  let bool;
  let newLetter = true;
  while (true) {
    try {
      while (check === 0) {
        if (newLetter) {
          stringsArray = Array.from(document.querySelectorAll('.music-headline-4'), el => el.textContent);
        }
        await sleep(1000);
        if (stringsArray.length > 1 && stringsArray[0] != undefined) {
          console.log("Epa! Encontramos uma letra... tentando romanizar");
          console.log(stringsArray);
          document.querySelectorAll('.music-headline-4').forEach(async (el) => {
            romanize(stringsArray[percursion])
              .then(res => {
                el.textContent = res;
                el.appendChild(document.createElement('br'));
              })
              .catch(error => {
                console.error('Erro na solicitação:', error);
              });
            percursion++;
          });
          check = 1;
          newLetter = true;
          console.log("Concluido!");
        }
      }

      await sleep(1000);
      console.log("Checando...");
      ArrayCheck = Array.from(document.querySelectorAll('.music-headline-4'), el => el.textContent);
      bool = ArrayCheck[5].includes(stringsArray[5]);
      console.log(stringsArray);
      console.log(ArrayCheck);
      if (stringsArray.length === ArrayCheck.length) {

      } else {
        check = 0;
      };
      console.log("Checando...2");
      if (ArrayCheck[0].includes(undefined)) {
        console.log("Err");
        check = 0;
        newLetter = false;
      }
    }
    catch (error) {
      console.log("Talvez essa musica não tem letra... ou esta passando de musica!");
    }
  }
}
console.log('Romazition for Amazon Music ver 0.1');
romanizeStrings();


Extension/script/popup.js
document.addEventListener("DOMContentLoaded", function () {
  var switchRomanization = document.getElementById("switchRomanization");
  var buttonStatus = true;
  switchRomanization.addEventListener("click", function () {
    if (buttonStatus === true) {
      switchRomanization.textContent = "Ativado";
    } else {
      switchRomanization.textContent = "Desativado";
    }
    chrome.tabs.query({ active: true, currentWindow: true }, function (tabs) {
      chrome.tabs.sendMessage(tabs[0].id, { action: "changeVariable" });
      buttonStatus = !buttonStatus;

      if (buttonStatus === true) {
        switchRomanization.textContent = "Ativado";
        chrome.storage.local.set({ romanizeIsEnable: true }, function () {
          console.log('Valor armazenado');
        });
      } else {
        switchRomanization.textContent = "Desativado";
        chrome.storage.local.set({ romanizeIsEnable: false }, function () {
          console.log('Valor armazenado');
        });
      }
    });
  });
});


Extension/script/romanization.js
let active = true;
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
function hasJapaneseCharacters(str) {
  var regex = /[\u3040-\u309F\u30A0-\u30FF\u31F0-\u31FF\u4E00-\u9FFF]/;
  return regex.test(str);
}

function romanize(dataIn) {
  const url = 'https://romanization.hirameki.me/';
  const inputData = 'input=' + dataIn;

  return fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded'
    },
    body: inputData
  })
    .then(response => response.text())
    .then(data => {
      return data;
    })
    .catch(error => {
      throw error;
    });
}
async function romanizeStrings() {
  let check = 0;
  let percursion = 0;
  let stringsArray;
  let ArrayCheck;
  let newLetter = true;
  while (true) {
    try {
      while (check === 0 && active) {
        if (newLetter) {
          stringsArray = Array.from(document.querySelectorAll('.music-headline-4'), el => el.textContent);
        }
        await sleep(500);
        if (stringsArray.length > 1 && stringsArray[0] != undefined && active) {
          console.log("Music is find, at valid letter, trying romanization");
          document.querySelectorAll('.music-headline-4').forEach(async (el) => {
            if (hasJapaneseCharacters(stringsArray[percursion])) {
              romanize(stringsArray[percursion])
                .then(res => {
                  el.appendChild(document.createElement('br'));
                  el.append(res+"​");
                })
                .catch(error => {
                  console.error('Error on API, please report', error);
                });
              percursion++;
            }else{
              el.append("​");
              percursion++;
            }
          });
          percursion = 0;
          check = 1;
          newLetter = true;
          console.log("Hey, this transliteration in this music ended!");
        }
      }
      await sleep(1000);
      ArrayCheck = Array.from(document.querySelectorAll('.music-headline-4'), el => el.textContent);
      if (ArrayCheck[0].includes("​")) {
      } else {
        check = 0;
      };
      if (ArrayCheck[0].includes(undefined)) {
        console.error("The transliteration falied, this erros mean that the script don't able a locate string on array");
        check = 0;
        newLetter = false;
      }
    }
    catch (error) {
      console.log("This music don't have letters... or music still loading... waiting...");
    }
  }
}

chrome.runtime.onMessage.addListener(function (request, sender, sendResponse) {
  if (request.action === "changeVariable") {
    active = !active;
  }
});
console.log('Romazition for Amazon Music ver 0.1');
romanizeStrings();


Extension/script/main.js
var active = true;
var percursion = 0;
var stringsArray = [];
var ArrayCheck = [];

chrome.runtime.onMessage.addListener(function (request, sender, sendResponse) {
  if (request.action === "changeVariable") {
    active = !active;
  }
});

function hasJapaneseCharacters(str) {
  var regex = /[\u3040-\u309F\u30A0-\u30FF\u31F0-\u31FF\u4E00-\u9FFF]/;
  return regex.test(str);
}

function romanize(dataIn) {
  const url = 'https://romanization.hirameki.me/';
  const inputData = 'input=' + dataIn;

  return fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded'
    },
    body: inputData
  })
    .then(response => response.text())
    .then(data => {
      return data;
    })
    .catch(error => {
      throw error;
    });
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function amazonMusic() {
  while (true) {
    try {
      await sleep(1000);
      while (active) {
        ArrayCheck = Array.from(document.querySelectorAll('.music-headline-4'), el => el.textContent);
        if (!ArrayCheck[0].includes("​")) {
          stringsArray = ArrayCheck;
          console.log("Music is find, at valid letter, trying romanization");
          document.querySelectorAll('.music-headline-4').forEach(async (el) => {
            if (hasJapaneseCharacters(stringsArray[percursion])) {
              romanize(stringsArray[percursion])
                .then(res => {
                  el.appendChild(document.createElement('br'));
                  el.append(res + "​");
                })
                .catch(error => {
                  console.error('Error on Put text, please report', error);
                });
              percursion++;
            } else {
              el.append("​");
              percursion++;
            }
          });
          percursion = 0;
          console.log("Hey, this transliteration in this music ended!");
        }
        await sleep(1000);
      }
    }
    catch (error) {
      console.log("This music don't have letters... or music still loading... waiting...");
    }
  }
}
async function spotify() {
  var elements;
  while (true) {
    await sleep(1000);
    try {
      while (active) {
        elements = document.querySelectorAll('[data-testid="fullscreen-lyric"]');
        ArrayCheck = [];
        elements.forEach(elemento => {
          ArrayCheck.push(elemento.textContent);
        });
        if (!ArrayCheck[0].includes("​")) {
          stringsArray = ArrayCheck;
          elements.forEach(element => {
            if (hasJapaneseCharacters(stringsArray[percursion])) {
              romanize(stringsArray[percursion])
                .then(res => {
                  element.appendChild(document.createElement('br'));
                  element.append(res + "​");
                  element.appendChild(document.createElement('br'));
                  element.appendChild(document.createElement('br'));
                })
                .catch(error => {
                  console.error('Error on Put text, please report', error);
                });
              percursion++;
            } else {
              element.append("​");
              element.appendChild(document.createElement('br'));
              element.appendChild(document.createElement('br'));
              percursion++;
            }
          });
          percursion = 0;
        }
        await sleep(1000);
      }
    }
    catch (error) {
      console.log("This music don't have letters... or music still loading... waiting...");
    }
  }
}
function init() {
  console.log('Romazition for Streaming Music Service ver 0.1... Detection Service...');
  var url = window.location.href;
  if (url.includes("amazon")) {
    console.log('Amazon Music Detected, starting romanization service on atrack in ', url);
    amazonMusic();
  } else if (url.includes("spotify")) {
    console.log('Spotify deceted, starting romanization service on atrack in ', url);
    spotify();
  }
}

init();

